File: libs/ui_composite/src/topology_checker_panel.cpp
#include <pwb/ui_composite/topology_checker_panel.hpp>

#include <pwb/ui_composite/composite_controller.hpp>
#include <pwb/ui_composite/topology_service.hpp>

#include <QColor>
#include <QFont>
#include <QHBoxLayout>
#include <QVariantMap>
#include <QVBoxLayout>

namespace pwb::ui_composite {

namespace {

const std::map<std::string, QString>& rule_labels() {
    static const std::map<std::string, QString> labels = {
        {"", QStringLiteral("全部规则")},
        {"overlap", QStringLiteral("面重叠")},
        {"gap", QStringLiteral("面缝隙")},
        {"is_valid", QStringLiteral("几何有效性")},
        {"workspace_remainder", QStringLiteral("工区余量")},
        {"dangle", QStringLiteral("线悬挂点")},
    };
    return labels;
}

std::string str_field(const Json& object, const char* key) {
    if (!object.is_object()) return "";
    auto it = object.find(key);
    if (it == object.end() || it->is_null()) return "";
    if (it->is_string()) return it->get<std::string>();
    if (it->is_number() || it->is_boolean()) return it->dump();
    return "";
}

QVariant json_to_variant(const Json& value) {
    if (value.is_null()) return {};
    if (value.is_boolean()) return value.get<bool>();
    if (value.is_number_integer()) {
        return QVariant::fromValue<qlonglong>(value.get<long long>());
    }
    if (value.is_number_unsigned()) {
        return QVariant::fromValue<qulonglong>(value.get<unsigned long long>());
    }
    if (value.is_number_float()) return value.get<double>();
    if (value.is_string()) {
        return QString::fromStdString(value.get<std::string>());
    }
    if (value.is_array()) {
        QVariantList list;
        for (const Json& entry : value) list.append(json_to_variant(entry));
        return list;
    }
    if (value.is_object()) {
        QVariantMap map;
        for (auto it = value.begin(); it != value.end(); ++it) {
            map[QString::fromStdString(it.key())] = json_to_variant(it.value());
        }
        return map;
    }
    return {};
}

QVariantMap json_to_map(const Json& object) {
    const QVariant variant = json_to_variant(object);
    return variant.toMap();
}

}  // namespace

TopologyCheckerPanel::TopologyCheckerPanel(QWidget* parent)
    : QWidget(parent) {
    setObjectName("TopologyCheckerPanel");

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(8, 8, 8, 8);

    badge = new QLabel(QStringLiteral("尚未检查"), this);
    badge->setObjectName("TopologyCheckerBadge");
    layout->addWidget(badge);

    rule_filter = new QComboBox(this);
    for (const auto& [key, label] : rule_labels()) {
        rule_filter->addItem(label, QString::fromStdString(key));
    }
    connect(rule_filter, &QComboBox::currentIndexChanged, this,
            [this](int) { rebuild(); });
    layout->addWidget(rule_filter);

    error_list = new QListWidget(this);
    error_list->setObjectName("TopologyCheckerErrorList");
    error_list->setContextMenuPolicy(Qt::CustomContextMenu);
    connect(error_list, &QListWidget::itemClicked, this,
            &TopologyCheckerPanel::on_item_clicked);
    connect(error_list, &QListWidget::customContextMenuRequested, this,
            &TopologyCheckerPanel::on_menu);
    layout->addWidget(error_list, 1);

    auto* buttons = new QHBoxLayout();
    check_button = new QPushButton(QStringLiteral("检查"), this);
    fix_button = new QPushButton(QStringLiteral("修复"), this);
    fix_all_button = new QPushButton(QStringLiteral("全部修复"), this);
    ignore_button = new QPushButton(QStringLiteral("忽略"), this);
    restore_button = new QPushButton(QStringLiteral("恢复"), this);
    connect(check_button, &QPushButton::clicked, this,
            &TopologyCheckerPanel::on_check);
    connect(fix_button, &QPushButton::clicked, this,
            &TopologyCheckerPanel::on_fix);
    connect(fix_all_button, &QPushButton::clicked, this,
            &TopologyCheckerPanel::fix_all_requested);
    connect(ignore_button, &QPushButton::clicked, this,
            &TopologyCheckerPanel::on_ignore);
    connect(restore_button, &QPushButton::clicked, this,
            &TopologyCheckerPanel::on_restore);
    for (QPushButton* button :
         {check_button, fix_button, fix_all_button, ignore_button,
          restore_button}) {
        buttons->addWidget(button);
    }
    layout->addLayout(buttons);
}

void TopologyCheckerPanel::bind(CompositeEditController* controller) {
    controller_ = controller;
}

void TopologyCheckerPanel::set_errors(
    const std::vector<Json>& errors,
    const std::set<IgnoreKey>& ignored_keys,
    const std::string& last_run_at) {
    all_errors_ = errors;
    ignored_keys_ = ignored_keys;
    if (!last_run_at.empty()) last_run_at_ = last_run_at;
    int blocking = 0;
    for (const Json& error : all_errors_) {
        if (!ignored_keys_.count(ignore_key(error))) ++blocking;
    }
    const QString stamp = QString::fromStdString(last_run_at_);
    if (!stamp.isEmpty()) {
        badge->setText(QStringLiteral("%1 处未忽略 · %2")
                           .arg(blocking)
                           .arg(stamp));
    } else if (!all_errors_.empty()) {
        badge->setText(QStringLiteral("%1 处未忽略").arg(blocking));
    } else {
        badge->setText(QStringLiteral("尚未检查"));
    }
    rebuild();
}

void TopologyCheckerPanel::rebuild() {
    const std::string rule =
        rule_filter->currentData().toString().toStdString();
    error_list->clear();
    for (const Json& error : all_errors_) {
        if (!rule.empty() && str_field(error, "rule") != rule) continue;
        QString label = QString::fromStdString(
            !str_field(error, "message").empty()
                ? str_field(error, "message")
                : !str_field(error, "rule").empty()
                      ? str_field(error, "rule")
                      : "error");
        const std::string feature_id = str_field(error, "feature_id");
        if (!feature_id.empty()) {
            label += QStringLiteral("（%1）")
                         .arg(QString::fromStdString(feature_id));
        }
        auto* item = new QListWidgetItem(label, error_list);
        item->setData(Qt::UserRole, json_to_variant(error));
        if (ignored_keys_.count(ignore_key(error))) {
            item->setForeground(QColor(160, 160, 160));
            QFont font = item->font();
            font.setItalic(true);
            item->setFont(font);
        }
    }
}

Json TopologyCheckerPanel::current_error() const {
    QListWidgetItem* item = error_list->currentItem();
    if (item == nullptr) return Json();
    const QVariant payload = item->data(Qt::UserRole);
    // QVariant -> Json round-trip is lossy for nested payloads; keep the
    // vector<Json> lookup by id instead.
    const QVariantMap map = payload.toMap();
    const std::string id = map.value("id").toString().toStdString();
    for (const Json& error : all_errors_) {
        if (str_field(error, "id") == id) return error;
    }
    // Fallback: reconstitute a shallow record (menus need id/methods only).
    Json out = Json::object();
    for (auto it = map.begin(); it != map.end(); ++it) {
        const QVariant v = it.value();
        if (v.typeId() == QMetaType::QString) {
            out[it.key().toStdString()] = v.toString().toStdString();
        } else if (v.typeId() == QMetaType::Double ||
                   v.typeId() == QMetaType::Int ||
                   v.typeId() == QMetaType::LongLong) {
            out[it.key().toStdString()] = v.toDouble();
        } else if (v.typeId() == QMetaType::Bool) {
            out[it.key().toStdString()] = v.toBool();
        }
    }
    return out;
}

void TopologyCheckerPanel::on_item_clicked(QListWidgetItem* item) {
    const Json error = current_error();
    (void)item;
    if (!error.is_object()) return;
    const Json bbox = error.value("bbox", Json::array());
    if (bbox.is_array() && bbox.size() >= 4) {
        emit zoom_requested({bbox[0].get<double>(), bbox[1].get<double>(),
                             bbox[2].get<double>(), bbox[3].get<double>()});
    }
    const std::string error_id = str_field(error, "id");
    if (!error_id.empty()) {
        emit highlight_requested(QString::fromStdString(error_id));
    }
}

void TopologyCheckerPanel::on_check() {
    if (controller_ != nullptr) {
        const std::vector<Json> errors = controller_->run_topology_checks();
        TopologyChecker& checker = controller_->topology().checker();
        std::string stamp = checker.last_run_at.value_or("");
        set_errors(errors, checker.ignored_keys(), stamp);
        return;
    }
    emit check_requested();
}

QMenu* TopologyCheckerPanel::menu_for(const Json& error) {
    auto* menu = new QMenu(this);
    Json methods = error.value("methods", Json::array());
    if (!methods.is_array() || methods.empty()) {
        methods = Json::array({Json{{"id", 0}, {"name", "修复"}}});
    }
    for (const Json& method : methods) {
        const std::string name = str_field(method, "name").empty()
                                     ? "修复"
                                     : str_field(method, "name");
        const std::string desc = str_field(method, "description");
        const QString label =
            desc.empty()
                ? QString::fromStdString(name)
                : QStringLiteral("%1 — %2")
                      .arg(QString::fromStdString(name),
                           QString::fromStdString(desc));
        QAction* action = menu->addAction(label);
        const int method_id = method.value("id", 0);
        const QString error_id =
            QString::fromStdString(str_field(error, "id"));
        connect(action, &QAction::triggered, this,
                [this, error_id, method_id]() {
                    emit fix_requested(error_id, method_id);
                });
    }
    return menu;
}

void TopologyCheckerPanel::on_menu(const QPoint& pos) {
    Json error = current_error();
    if (!error.is_object()) {
        QListWidgetItem* item = error_list->itemAt(pos);
        if (item != nullptr) {
            error_list->setCurrentItem(item);
            error = current_error();
        }
    }
    if (!error.is_object()) return;
    menu_for(error)->exec(error_list->viewport()->mapToGlobal(pos));
}

void TopologyCheckerPanel::on_fix() {
    const Json error = current_error();
    if (!error.is_object()) return;
    Json methods = error.value("methods", Json::array());
    const int method_id = methods.is_array() && !methods.empty()
                              ? methods[0].value("id", 0)
                              : 0;
    emit fix_requested(QString::fromStdString(str_field(error, "id")),
                       method_id);
}

void TopologyCheckerPanel::on_ignore() {
    const Json error = current_error();
    if (error.is_object()) emit ignore_requested(json_to_map(error));
}

void TopologyCheckerPanel::on_restore() {
    const Json error = current_error();
    if (error.is_object()) emit restore_requested(json_to_map(error));
}

}  // namespace pwb::ui_composite

File: legacy/python_reference/product/paleo_workbench/ui/workstation/topology_checker_panel.py
"""拓扑检查器面板（拓扑编辑迁移 M4 §5）。

错误列表、规则过滤、点击缩放/高亮、忽略灰显可恢复、单条/全部修复。
不嵌 QgsGeometryCheckerDialog（依赖 QgisInterface）。轻量 QWidget，
可作检查器 tab 或独立测。
"""
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtGui import QColor, QFont
from PySide6.QtWidgets import (
    QComboBox,
    QHBoxLayout,
    QLabel,
    QListWidget,
    QListWidgetItem,
    QMenu,
    QPushButton,
    QVBoxLayout,
    QWidget,
)

from paleo_workbench.mapping.topology_checker import ignore_key

__all__ = ["TopologyCheckerPanel"]

_RULE_LABELS = {
    "": "全部规则",
    "overlap": "面重叠",
    "gap": "面缝隙",
    "is_valid": "几何有效性",
    "workspace_remainder": "工区余量",
    "dangle": "线悬挂点",
}


class TopologyCheckerPanel(QWidget):
    """错误列表 + 检查/修复/忽略。"""

    zoom_requested = Signal(list)  # bbox [xmin, ymin, xmax, ymax]
    highlight_requested = Signal(str)  # error id
    check_requested = Signal()
    fix_requested = Signal(str, int)  # error id, method
    fix_all_requested = Signal()
    ignore_requested = Signal(dict)
    restore_requested = Signal(dict)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("TopologyCheckerPanel")
        self._all_errors: list[dict] = []
        self._ignored_keys: set[tuple] = set()
        self._controller = None
        self._last_run_at = None

        layout = QVBoxLayout(self)
        layout.setContentsMargins(8, 8, 8, 8)

        self.badge = QLabel("尚未检查", self)
        self.badge.setObjectName("TopologyCheckerBadge")
        layout.addWidget(self.badge)

        self.rule_filter = QComboBox(self)
        for key, label in _RULE_LABELS.items():
            self.rule_filter.addItem(label, key)
        self.rule_filter.currentIndexChanged.connect(self._rebuild)
        layout.addWidget(self.rule_filter)

        self.error_list = QListWidget(self)
        self.error_list.setObjectName("TopologyCheckerErrorList")
        self.error_list.setContextMenuPolicy(
            Qt.ContextMenuPolicy.CustomContextMenu)
        self.error_list.itemClicked.connect(self._on_item_clicked)
        self.error_list.customContextMenuRequested.connect(self._on_menu)
        layout.addWidget(self.error_list, 1)

        buttons = QHBoxLayout()
        self.check_button = QPushButton("检查")
        self.fix_button = QPushButton("修复")
        self.fix_all_button = QPushButton("全部修复")
        self.ignore_button = QPushButton("忽略")
        self.restore_button = QPushButton("恢复")
        self.check_button.clicked.connect(self._on_check)
        self.fix_button.clicked.connect(self._on_fix)
        self.fix_all_button.clicked.connect(self.fix_all_requested.emit)
        self.ignore_button.clicked.connect(self._on_ignore)
        self.restore_button.clicked.connect(self._on_restore)
        for button in (self.check_button, self.fix_button, self.fix_all_button,
                       self.ignore_button, self.restore_button):
            buttons.addWidget(button)
        layout.addLayout(buttons)

    def bind(self, controller) -> None:
        """绑定综合编修控制器（检查/修复走其 TopologyChecker）。"""
        self._controller = controller

    def set_errors(self, errors, ignored_keys=(), last_run_at=None) -> None:
        self._all_errors = [dict(error) for error in errors or ()]
        self._ignored_keys = set(ignored_keys or ())
        if last_run_at:
            self._last_run_at = last_run_at
        blocking = sum(1 for error in self._all_errors
                       if ignore_key(error) not in self._ignored_keys)
        stamp = str(self._last_run_at or "")
        if stamp:
            self.badge.setText(f"{blocking} 处未忽略 · {stamp}")
        elif self._all_errors:
            self.badge.setText(f"{blocking} 处未忽略")
        else:
            self.badge.setText("尚未检查")
        self._rebuild()

    def _rebuild(self) -> None:
        rule = str(self.rule_filter.currentData() or "")
        self.error_list.clear()
        for error in self._all_errors:
            if rule and str(error.get("rule") or "") != rule:
                continue
            label = str(error.get("message") or error.get("rule") or "error")
            feature_id = str(error.get("feature_id") or "")
            if feature_id:
                label = f"{label}（{feature_id}）"
            item = QListWidgetItem(label)
            item.setData(Qt.ItemDataRole.UserRole, dict(error))
            if ignore_key(error) in self._ignored_keys:
                item.setForeground(QColor(160, 160, 160))
                font = QFont(item.font())
                font.setItalic(True)
                item.setFont(font)
            self.error_list.addItem(item)

    def _current_error(self) -> dict | None:
        item = self.error_list.currentItem()
        if item is None:
            return None
        payload = item.data(Qt.ItemDataRole.UserRole)
        return dict(payload) if isinstance(payload, dict) else None

    def _on_item_clicked(self, item: QListWidgetItem) -> None:
        error = item.data(Qt.ItemDataRole.UserRole)
        if not isinstance(error, dict):
            return
        bbox = error.get("bbox") or []
        if len(bbox) >= 4:
            self.zoom_requested.emit([float(bbox[0]), float(bbox[1]),
                                      float(bbox[2]), float(bbox[3])])
        error_id = str(error.get("id") or "")
        if error_id:
            self.highlight_requested.emit(error_id)

    def _on_check(self) -> None:
        controller = self._controller
        if controller is not None and hasattr(controller, "run_topology_checks"):
            errors = controller.run_topology_checks()
            checker = controller.topology.checker
            self.set_errors(errors, ignored_keys=checker.ignored_keys(),
                            last_run_at=checker.last_run_at)
            return
        self.check_requested.emit()

    def menu_for(self, error: dict) -> QMenu:
        """单条右键：check 声明的方法列表（含预览描述）。"""
        menu = QMenu(self)
        methods = error.get("methods") or [{"id": 0, "name": "修复"}]
        for method in methods:
            name = str(method.get("name") or "修复")
            desc = str(method.get("description") or "")
            label = f"{name} — {desc}" if desc else name
            action = menu.addAction(label)
            method_id = int(method.get("id") or 0)
            error_id = str(error.get("id") or "")
            action.triggered.connect(
                lambda *_, eid=error_id, mid=method_id:
                    self.fix_requested.emit(eid, mid))
        return menu

    def _on_menu(self, pos) -> None:
        error = self._current_error()
        if error is None:
            item = self.error_list.itemAt(pos)
            if item is not None:
                self.error_list.setCurrentItem(item)
                payload = item.data(Qt.ItemDataRole.UserRole)
                error = dict(payload) if isinstance(payload, dict) else None
        if error is None:
            return
        self.menu_for(error).exec(self.error_list.mapToGlobal(pos))

    def _on_fix(self) -> None:
        error = self._current_error()
        if error is None:
            return
        methods = error.get("methods") or [{"id": 0}]
        self.fix_requested.emit(
            str(error.get("id") or ""), int(methods[0].get("id") or 0))

    def _on_ignore(self) -> None:
        error = self._current_error()
        if error is not None:
            self.ignore_requested.emit(error)

    def _on_restore(self) -> None:
        error = self._current_error()
        if error is not None:
            self.restore_requested.emit(error)
